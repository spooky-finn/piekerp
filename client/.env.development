REACT_APP_NODE_ENV=development

REACT_APP_API_URL=http://localhost:9000/api
REACT_APP_HASURA_ENDPOINT=reserve.hasura.piek.ru/v1/graphql