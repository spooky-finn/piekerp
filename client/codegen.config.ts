import { CodegenConfig } from '@graphql-codegen/cli'
import * as dotenv from 'dotenv'
dotenv.config({ path: `../.env` })

if (!process.env.HASURA_ENDPOINT || !process.env.HASURA_ADMIN_SECRET) {
  throw Error('Hasura connection params must be provided (☆^ー^☆)')
}

const config: CodegenConfig = {
  schema: [
    {
      [process.env.HASURA_ENDPOINT]: {
        headers: {
          'x-hasura-admin-secret': process.env.HASURA_ADMIN_SECRET
        }
      }
    }
  ],
  documents: ['./**/*.gql'],
  generates: {
    './src/types/graphql-shema.ts': {
      plugins: ['typescript', 'typescript-operations', 'typescript-react-apollo']
    }
  }
}

export default config
