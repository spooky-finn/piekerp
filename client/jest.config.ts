/** @type {import('ts-jest').JestConfigWithTsJest} */
import type { Config } from '@jest/types'

const config: Config.InitialOptions = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  setupFilesAfterEnv: ['<rootDir>/setUpTests.ts'],

  moduleNameMapper: {
    '\\.svg': '<rootDir>/__mocks__/svgrMock.tsx',

    '.+\\.(css|styl|less|sass|scss)$': '<rootDir>/node_modules/identity-obj-proxy',

    '\\.(jpg|jpeg|png|gif|eot|otf|webp|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
      '<rootDir>/__mocks__/fileMock.ts'
  }
}
export default config
