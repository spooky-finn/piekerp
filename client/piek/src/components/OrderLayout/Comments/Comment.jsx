import sass from './index.module.sass'
import moment from 'moment'

const Comment = (props) => {
  const { data, userID, updateComment, deleteComment, nowEditing, setNowEditing } = props

  function sender(){
    if (userID === data.User.UserID) return 'Вы'
    else return `${data.User.FirstName} ${data.User.LastName}`
  }
  function actions(){
    if (userID === data.User.UserID && nowEditing?.CommentID !== data.CommentID ) return (
        <div onClick={()=> deleteComment(data.CommentID)}>Удалить</div>
    )
  }
  function timestamp(){
    const date = moment(data.Timestamp)
    return date.format('MMM D') +' at '+ date.format('hh:mm')
  }

  function getCommentContent(){
    if (userID === data.User.UserID) return <div contentEditable="true"
    key={data.CommentID}
    className={sass.commentContent}
    suppressContentEditableWarning="true"
    onBlur={async (e) => {
      await updateComment(data.CommentID, e.target.innerHTML)
    }}
    onMouseLeave={async (e) => {
      await updateComment(data.CommentID, e.target.innerHTML)
      setNowEditing(null)
    }}
    dangerouslySetInnerHTML={{ __html: data.Text}}></div>

    else return <div className={sass.commentContent}>{data.Text}</div>

  }

  return (
    <div className={sass.commentUnit}>

      <div className={sass.commentHeader}>
        <div className={sass.sender}> {sender()} </div>
        <div className={sass.actions}>{actions()}</div>
        <div className={sass.time}> {timestamp()} </div>
      </div>
      {getCommentContent()}

    </div>
  )
}
export default Comment 