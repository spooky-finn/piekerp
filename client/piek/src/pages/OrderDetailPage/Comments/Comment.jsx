import sass from './index.module.sass'
import moment from 'moment'

const Comment = (props) => {
    const { data, userID, updateComment, deleteComment, nowEditing, setNowEditing } = props

    function sender(){
        return `${data.User.FirstName} ${data.User.LastName}`
    }
    
    function actions(){
        if (userID === data.User.UserID && nowEditing?.CommentID !== data.CommentID ) return (
            <div onClick={()=> deleteComment(data.CommentID)}>Удалить</div>
        )
    }
    function timestamp(){
        const date = moment(data.Timestamp)
        return date.format('MMM D') +' at '+ date.format('hh:mm')
    }

    function changeInCommentTodo(e){
        // если комментарий содержит чек-лист выполняем мутацию на обновление состояния
        if (e.currentTarget.innerHTML.toString().includes(sass.checklistUnit, 0)){
            updateComment(data.CommentID, e.currentTarget.innerHTML)
        }

    }
    
    function getCommentContent(){
        const isYourComment = (userID === data.User.UserID)
        if (isYourComment) return (
            <div 
                contentEditable = "true"
                key             = {data.CommentID}
                className       = {sass.commentContent}
                suppressContentEditableWarning = "true"
                onBlur={async (e) => {
                await updateComment(data.CommentID, e.target.innerHTML)
                }}
                onMouseLeave={async (e) => {
                await updateComment(data.CommentID, e.target.innerHTML)
                setNowEditing(null)
                }}
                onMouseDown = {((e) => {
                    // execute content selecting only on doubleclick or over
                        if (e.detail < 2) e.preventDefault();                       
                    })}
                dangerouslySetInnerHTML={{ __html: data.Text}}>
            </div>)

        else return (
            <div 
                onClick={changeInCommentTodo}
                dangerouslySetInnerHTML={{ __html: data.Text }} 
                className={sass.commentContent}>
            </div>
        )

    }

    return (
        <div className={sass.commentUnit}>

        <div className={sass.commentHeader}>
            <div className={sass.sender}> {sender()} </div>
            <div className={sass.actions}>{actions()}</div>
            <div className={sass.time}> {timestamp()} </div>
        </div>
        {getCommentContent()}

        </div>
    )
}
export default Comment 