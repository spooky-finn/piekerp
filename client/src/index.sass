p, span, div, li, td, tr, th, input
  color: var(--highContrast)

html
  background: var(--L0) !important
  font-size: 14px
  @media (max-width: 1000px)
    font-size: .9rem

a 
  text-decoration: none
  &:hover 
    color: var(--accent)
    
body
  margin: 0
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',sans-serif
  -webkit-font-smoothing: antialiased
  -moz-osx-font-smoothing: grayscale
  overflow: hidden


/* Chrome, Safari, Edge, Opera */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button 
  -webkit-appearance: none
  margin: 0


/* Firefox */
input[type=number] 
  -moz-appearance: textfield


input:active, input:focus
  outline: none !important


[contentEditable=true]:empty:before
  content: attr(data-ph)
  color: var(--lowContrast)
  opacity: .5

@media print
  .noprint
    display: none !important

.App
  display: grid
  grid-template-columns: 51px auto
  height: 100vh
  @media print
    height: auto !important

.reactTooltips
  font-size: .8
  padding: 5px 8px !important



.pageLayout__header
  display: flex
  align-items: center
  padding: 10px 15px 15px
  min-height: 34px
  font-size: 1.1rem
  .pageLayout__title
    font-weight: 600
    color: var(--accent)
  .pageLayout__icon
    margin-right: .5rem
    width: 23px
    path
      color: var(--accent)
  .pageLayout__actions
    margin-left: auto

