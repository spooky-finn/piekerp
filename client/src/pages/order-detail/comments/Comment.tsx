import moment from 'moment'
import { TComment } from 'src/types/global'
import { useDeleteCommentMutation, useUpdateCommentMutation } from 'src/types/graphql-shema'
import sass from './index.module.sass'

interface ICommentProps {
  data: TComment
  userID: number
}

export default function Comment({ data, userID }: ICommentProps) {
  const [deleteMutation] = useDeleteCommentMutation()
  const [updateMutation] = useUpdateCommentMutation()

  function updateComment(id: number, newText: string) {
    updateMutation({
      variables: {
        CommentID: id,
        Text: newText.trim()
      }
    })
  }

  function handleDelete() {
    deleteMutation({ variables: { CommentID: data.CommentID } })
  }

  const handleChangeTodo: React.MouseEventHandler<HTMLDivElement> = e => {
    // если комментарий содержит чек-лист выполняем мутацию на обновление состояния
    if (e.currentTarget.innerHTML.toString().includes(sass.checklistUnit, 0)) {
      updateComment(data.CommentID, e.currentTarget.innerHTML)
    }
  }

  function sender() {
    return `${data.User.FirstName} ${data.User.LastName}`
  }

  function actions() {
    if (userID === data.User.UserID) return <div onClick={handleDelete}>Удалить</div>
  }

  function timestamp() {
    const date = moment(data.Timestamp)
    return date.format('MMM D') + ' at ' + date.format('hh:mm')
  }

  function getCommentContent() {
    const isYourComment = userID === data.User.UserID
    return isYourComment ? (
      <div
        contentEditable="true"
        key={data.CommentID}
        className={sass.commentContent}
        suppressContentEditableWarning={true}
        onBlur={async e => {
          await updateComment(data.CommentID, e.target.innerHTML)
        }}
        onMouseLeave={async e => {
          await updateComment(data.CommentID, e.currentTarget.innerHTML)
          // setNowEditing(null)
        }}
        dangerouslySetInnerHTML={{ __html: data.Text }}
      />
    ) : (
      <div
        onClick={handleChangeTodo}
        dangerouslySetInnerHTML={{ __html: data.Text }}
        className={sass.commentContent}
      />
    )
  }

  return (
    <div className={sass.commentUnit}>
      <div className={sass.commentHeader}>
        <div className={sass.sender}> {sender()} </div>
        <div className={sass.actions}>{actions()}</div>
        <div className={sass.time}> {timestamp()} </div>
      </div>
      {getCommentContent()}
    </div>
  )
}
