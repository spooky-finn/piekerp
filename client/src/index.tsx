import { ApolloProvider } from '@apollo/client'
import { createContext } from 'react'
import { createRoot } from 'react-dom/client'
import { BrowserRouter } from 'react-router-dom'
import { apolloClient } from './api/apollo/apollo-client'
import App from './App'
import './index.sass'
import './material/theme.css'
import Store from './store/store'

const store = new Store()

export const Context = createContext({
  store
})

export function getInMemoryToken() {
  return store.inMemoryToken
}

export async function getNewInMemoryToken() {
  return await store.getNewToken()
}

const container = document.getElementById('app') || document.createElement('div')
const root = createRoot(container) // createRoot(container!) if you use TypeScript

root.render(
  <ApolloProvider client={apolloClient}>
    <BrowserRouter>
      <Context.Provider value={{ store }}>
        <App />
      </Context.Provider>
    </BrowserRouter>
  </ApolloProvider>
)
