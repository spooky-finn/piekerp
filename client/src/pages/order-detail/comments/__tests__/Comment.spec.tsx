import { MockedProvider } from '@apollo/client/testing'
import { render, screen } from '@testing-library/react'
import UserEvent from '@testing-library/user-event/'
import Comment from '../Comment'

function setup(userID: number) {
  const { asFragment } = render(
    <MockedProvider>
      <Comment
        userID={userID}
        data={{
          CommentID: 112,
          Text: 'there is a comment',
          Timestamp: '2021-11-24T07:42:13.342852+00:00',
          User: {
            UserID: 15,
            FirstName: 'Fin',
            LastName: 'Kelstein'
          }
        }}
      />
    </MockedProvider>
  )
  return {
    asFragment
  }
}

test('snapshot test', () => {
  const { asFragment } = setup(15)
  expect(asFragment()).toMatchSnapshot()
})

test('should show delete button if user is the owner of comment', () => {
  setup(15)
  expect(screen.getByText('Удалить')).toBeInTheDocument()
})

test('should not display delete Btn', () => {
  setup(12)
  expect(screen.queryByText('Удалить')).not.toBeInTheDocument()
})

test('should be able for the owner to input something in content field', async () => {
  setup(15)
  const contentFiled = screen.getByText('there is a comment') as HTMLDivElement
  await UserEvent.type(contentFiled, 'something new')
  expect(contentFiled.textContent).toContain('something new')
})
