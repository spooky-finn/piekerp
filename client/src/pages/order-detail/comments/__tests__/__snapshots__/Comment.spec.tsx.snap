// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<DocumentFragment>
  <div
    class="commentUnit"
  >
    <div
      class="commentHeader"
    >
      <div
        class="sender"
      >
         Fin Kelstein 
      </div>
      <div
        class="actions"
      >
        <div>
          Удалить
        </div>
      </div>
      <div
        class="time"
      >
         Nov 24 at 10:42 
      </div>
    </div>
    <div
      class="commentContent"
      contenteditable="true"
    >
      there is a comment
    </div>
  </div>
</DocumentFragment>
`;
